docs/proposals/gpu-support.md

```diff
@@ -0,0 +1,339 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN MUNGE: GENERATED_TOC -->
+
+- [GPU support](#gpu-support)
+  - [Objective](#objective)
+  - [Background](#background)
+  - [Detailed discussion](#detailed-discussion)
+    - [Inventory](#inventory)
+    - [Scheduling](#scheduling)
+    - [The runtime](#the-runtime)
+      - [NVIDIA support](#nvidia-support)
+    - [Event flow](#event-flow)
+  - [Implementation plan](#implementation-plan)
+    - [V0](#v0)
+      - [Scheduling](#scheduling)
+      - [Runtime](#runtime)
+      - [Other](#other)
+  - [Future work](#future-work)
+    - [V1](#v1)
+    - [V2](#v2)
+    - [V3](#v3)
+    - [Undetermined](#undetermined)
+  - [Security considerations](#security-considerations)
+
+<!-- END MUNGE: GENERATED_TOC -->
+
+# GPU support
+
+Author: @therc
+
+Date: Apr 2016
+
+Status: Design in progress, early implementation of requirements
+
+## Objective
+
+Users should be able to request GPU resources for their workloads, as easily as
+for CPU or memory. Kubernetes should keep an inventory of machines with GPU
```

Comment:
@davidopp in another thread suggested to describe this kind non-fungible resource using a different approach called opaque counted resource at: https://github.com/kubernetes/kubernetes/issues/19049#issuecomment-170292193

I think that approach can cover the three attributes listed above by @timothysc. So that Kubelet can expose something like:

GPU-Version+Driver-Version: count through NodeCapacity / NodeAllocatable. 

Meanwhile, we also talked about this in today's sig-node meeting on the initial support (v0) with more restriction: 
* Only allow one GPU per node
* Introducing an experimental flag which includes GPU version info, and hardcode Kubelet to report GPU as NodeCapacity & NodeAllocatable if the flag is on. Also report this information through node label.
* Do we need to have a flag to config host device associating with GPU flag? I forgot our decision on this one.
* In this approach, required libraries have to be included in image itself. A given pod requesting a GPU, Kubelet based on the configuration to setup the runtime environment. 

cc/ @erictune Did I miss anything above? 

Once initContainer or VolumeContainer work is done, we can use that feature to allow preconfiged nvidia-docker volume driver, and Kubelet can discover those, and expose those using opaque counted resource approach above.